sdv/cag/range.py

```diff
@@ -0,0 +1,376 @@
+        mid_column_sdtype = metadata.tables[table_name].columns[self._middle_column_name]['sdtype']
+        high_column_sdtype = metadata.tables[table_name].columns[self._high_column_name]['sdtype']
+        if low_column_sdtype != mid_column_sdtype or mid_column_sdtype != high_column_sdtype:
+            raise PatternNotMetError(
+                f"Columns '{self._low_column_name}', '{self._middle_column_name}' and "
+                f"'{self._high_column_name}' must have the same sdtype. Found '{low_column_sdtype}'"
+                f", '{mid_column_sdtype}' and '{high_column_sdtype}'."
+            )
+
+    def _get_data(self, data):
+        low = data[self._low_column_name].to_numpy()
+        mid = data[self._middle_column_name].to_numpy()
+        high = data[self._high_column_name].to_numpy()
+        return low, mid, high
+
+    def _validate_pattern_with_data(self, data, metadata):
+        """Validate the data is compatible with the pattern."""
+        table_name = self._get_single_table_name(metadata)
+        data = data[table_name]
+        low, mid, high = self._get_data(data)
+
+        low_is_nan = pd.isna(low)
+        mid_is_nan = pd.isna(mid)
+        high_is_nan = pd.isna(high)
+
+        low_lt_mid = self._operator(low, mid) | low_is_nan | mid_is_nan
+        mid_lt_high = self._operator(mid, high) | mid_is_nan | high_is_nan
+        low_lt_high = self._operator(low, high) | low_is_nan | high_is_nan
+
+        valid = low_lt_mid & mid_lt_high & low_lt_high
+
+        if not valid.all():
+            invalid_rows = np.where(~valid)[0]
+            if len(invalid_rows) <= 5:
+                invalid_rows_str = ', '.join(str(i) for i in invalid_rows)
+            else:
+                first_five = ', '.join(str(i) for i in invalid_rows[:5])
+                remaining = len(invalid_rows) - 5
+                invalid_rows_str = f'{first_five}, +{remaining} more'
+
+            raise PatternNotMetError(
+                f'The range requirement is not met for row indices: [{invalid_rows_str}]'
+            )
+
+    def _get_updated_metadata(self, metadata):
+        """Get the new output metadata after applying the pattern to the input metadata."""
+        table_name = self._get_single_table_name(metadata)
+        low_diff_column = _create_unique_name(
+            self._low_diff_column_name, metadata.tables[table_name].columns.keys()
+        )
+        high_diff_column = _create_unique_name(
+            self._high_diff_column_name, metadata.tables[table_name].columns.keys()
+        )
+
+        metadata = metadata.to_dict()
+        metadata['tables'][table_name]['columns'][low_diff_column] = {'sdtype': 'numerical'}
+        metadata['tables'][table_name]['columns'][high_diff_column] = {'sdtype': 'numerical'}
+        del metadata['tables'][table_name]['columns'][self._high_column_name]
+        del metadata['tables'][table_name]['columns'][self._middle_column_name]
```

Comment:
The issue says to drop middle_column_name and high_column_name, but I wonder why we would not drop low_column_name

Reply:
Both `Range` and `Inequality` constraint/cag create a `diff` column during transform. When reversing it adds the `diff` column to the `low` column to get back the middle/high columns. Both constraints need the `low` column for this process, so it's not removed from the data/metadata.